### PR TITLE
feat: add pagination support

### PR DESCRIPTION
**Pull request by jerome-quere**

### CODE REVIEW

File: scw/request.go [hunk 1]
```diff
@@ -66,17 +74,47 @@ func (c *Client) Do(req *ScalewayRequest, opts ...RequestOption) (*http.Response
...
+	for i := uint64(0); i < total; i++ {
+		r := reflect.New(reflect.ValueOf(res).Type())
+		req.Pagination.Page = 1      // Init pagination
+		err := c.Do(req, r, opts...) // Do first request
+		if err != nil {
+			return err
+		}
+		slice = append(slice, reflect.ValueOf(res).FieldByName("????").Interface().([]interface{})...)
+	}
+	reflect.ValueOf(res).FieldByName("????").Set(reflect.ValueOf(slice))
```
Comment (kindermoumoute) [inline scw/request.go]: This is where and interface would be powerful:
```Go
type lister interface {
    Results() []interface{}
    Append([]interface{})
}
//...

switch v := res.(type) {
case lister:
	// do first request to get total
	
	var results []interface{}
	req.Pagination.Page = 0      // Init pagination
	for i := uint64(0); i < total; i+=len(results) {
		r := reflect.New(reflect.ValueOf(res).Type())
		req.Pagination.Page++
		err := c.Do(req, r, opts...) // Do first request
		if err != nil {
			return err
		}
		results = r.(lister).Results()
		v.Append(results)
	}
default:
    return fmt.Errorf("pagination is not supported on %s", v)
}
```

File: scw/client.go [hunk 2]
```diff
@@ -90,6 +91,10 @@ func defaultOptions() []ClientOption {
 	return []ClientOption{
 		WithAPIURL("https://api.scaleway.com"),
 		WithUserAgent(userAgent),
+		WithPaginationConfig(&PaginationConfig{
```
Comment (kindermoumoute) [inline scw/client.go]: IMO clientOption must have static parameters:

```Go
WithPaginationConfig(full bool, pageSize uint)
```

Let's decide fully decide what this client option is about so we fix these parameters for good.

File: scw/client_option.go [hunk 3]
```diff
@@ -121,3 +121,14 @@ func WithDefaultZone(zone utils.Zone) ClientOption {
 		s.defaultZone = zone
 	}
 }
+
+func WithPaginationConfig(paginationConfig *PaginationConfig) ClientOption {
```
Comment (kindermoumoute) [inline scw/client_option.go]: `WithPagination()`

File: scw/client.go [hunk 4]
```diff
@@ -90,6 +91,10 @@ func defaultOptions() []ClientOption {
...
+			Full:     false,
```
Comment (QuentinBrosse) [inline scw/client.go]: What is the exact purpose of `full`? If it's true the pagination is disabled?
Maybe we could find a better name

File: scw/client_option.go [hunk 5]
```diff
@@ -121,3 +121,14 @@ func WithDefaultZone(zone utils.Zone) ClientOption {
...
+	return func(s *settings) {
+		s.paginationConfig = *paginationConfig
+	}
+}
+
+func WithoutPagination() ClientOption {
```
Comment (kindermoumoute) [inline scw/client_option.go]: We should achieve "no pagination" (what does it mean?) with `WithPagination`

**Pull request closed without merging**